### DISCUSSION TITLE
Custom encryption decryption

### DISCUSSION THREAD

**Discussion by D00mSlayer** [sqlalchemy/sqlalchemy#11755] [Usage Questions]
I am using MS SQL with SQLAlchemy. All my tables have integer primary keys, and I now have a requirement to have these encrypted so that these IDs are not exposed to the UI.

This is what I have tried so far.
I want to be able to always fetch the encrypted pk and even the original pk (based on workflows). 

```
import os
import base64
import pyodbc
import sqlalchemy as sa
from sqlalchemy.orm import sessionmaker
from z3c.sqlalchemy.mapper import MappedClassBase
from sqlalchemy.orm import object_session
from sqlalchemy import MetaData
from sqlalchemy.engine.base import Engine
from sqlalchemy.ext.declarative import declarative_base
from cryptography.fernet import Fernet
from sqlalchemy.ext.hybrid import hybrid_property


class _EntityBase(MappedClassBase):
    @property
    def session(self):
        return object_session(self)

    def query(self, *args, **argd):
        return self.session.query(*args, **argd)

common_sqla_metadata = MetaData()
DbEntity = declarative_base(name='DbEntity', metadata=common_sqla_metadata, cls=_EntityBase)

rand = os.urandom(32)
KEY = base64.urlsafe_b64encode(rand)

def encrypt(plainint):
    fernet = Fernet(KEY)
    return fernet.encrypt(str(plainint).encode()).decode()

def decrypt(ciphertext):
    fernet = Fernet(KEY)
    return int(fernet.decrypt(ciphertext.encode()).decode())


class EncryptedPK(sa.TypeDecorator):
    impl = sa.BigInteger

    def process_bind_param(self, value, dialect):
        return value

    def process_result_value(self, value, dialect):
        return value


class TableA(DbEntity):
    __tablename__ = 'table_a'

    _table_a_id = sa.Column("table_a_id", sa.BigInteger, primary_key=True)
    name = sa.Column("name", sa.String(200, "SQL_Latin1_General_CP1_CI_AS"))

    @hybrid_property
    def table_a_id(self):
        return encrypt(self._table_a_id)

    @table_a_id.expression
    def table_a_id(cls):
        return sa.func.dbo.EncryptData2(cls._table_a_id, KEY)


class TableB(DbEntity):
    __tablename__ = 'table_b'

    table_b_id = sa.Column("table_b_id", sa.BigInteger, primary_key=True)
    name = sa.Column("name", sa.String(200, "SQL_Latin1_General_CP1_CI_AS"))
    table_a_id = sa.Column(sa.ForeignKey('table_a.table_a_id'))


e = sa.create_engine(
    "mssql+pyodbc://sa:tiger^5HHH@127.0.0.1:1433/test?driver=ODBC+Driver+18+for+SQL+Server&TrustServerCertificate=yes",
    echo=False,
)
Session = sessionmaker(bind=e)
session = Session()

DbEntity.metadata.drop_all(session.get_bind())
DbEntity.metadata.create_all(session.get_bind())

entity = TableA(name="qweerty")
session.add(entity)
session.commit()
e_pk = entity.table_a_id
pk = entity._table_a_id
print("===== INSERTED - ENCRYPTED =====:", e_pk)
print("===== INSERTED - DECRYPTED =====:", pk)
# import pdb; pdb.set_trace()
res = session.query(
    TableA.table_a_id.label('encrypted_id'),
    TableA._table_a_id.label('original_id')
).filter(
    TableA._table_a_id == pk
).all()
print(res)
```

EncryptData2 is a custom function that I tried to create in SQL Server in the hopes that it would work the same way as in python. However, this isn't the case.

I tried a custom typedecoration with EncryptedPK but no luck with this either.

Considering that my product is huge, I cannot afford a lot of code changes. I am okay with changing models or even some queries and business logic (when a workflow requires the original pk). I want:
1. an option to get the original and encrypted pks after inserting.
2. an option to select the original and encrypted pks in a query
3. an option to automatically decrypt (if encrypted) these values when passed to a query

Any help would be appreciated.
Thanks

**Comment by zzzeek** [comment]
hi - 

for client side encrpytion with hybrid property take a look at the canonical example at https://github.com/sqlalchemy/sqlalchemy/wiki/SymmetricEncryptionClientSide , which uses a custom Comparator to correctly define using encryption on the compared value, so that you compare encrypted strings on the database side.   note that the encryption scheme must produce identical encryption strings each time for this comparison to work.

**Reply by D00mSlayer** [reply]
@zzzeek we would want to avoid client side encryption / decryption.
I am looking for a solution that will:
1. encrypt a BigInteger primary key column on fetch (either entire row or that specific column)
2. decrypt the encrypted value during bind params back to original BigInteger primary key
3. also have an ability to fetch the original BigInteger PK when using a select statement.

I tried a solution with a combination of a custom TypeDecorator with hybrid_property but I am unable to fetch the original value when the attribute is used in a select clause.

**Reply by zzzeek** [reply]
your example script shows the use of a client side encrypt/decrypt function.   by "we want to avoid", do you mean that you want to use a server-side encrypt / decrypt function?  or that you want to continue using client side encrypt/decrypt, however you wish to minimize the use of these functions?    The example I linked in the wiki illustrates the client side approach.  if you are trying to SELECT values from a database using WHERE criteria that compares to an encrypted column, you either need to encrypt the client side value to compare, or decrypt on the server side in your WHERE criteria.   

Looking at your example again it looks like you have *both* a client side funciton *and* what may be some server side function present.   if you have server side functions available, you can use those alone, see the server side example: https://github.com/sqlalchemy/sqlalchemy/wiki/SymmetricEncryptionServerSide .  the second example shows the use of a TypeDecorator.

we do not currently have an example that mixes client and server side encryption together as this is typically not needed.

**Reply by D00mSlayer** [reply]
There is no need for client side encryption. we are okay to display encrypted keys in the UI. These values when sent back to the server need to be decrypted to their original ID when bound as params to a query. There are some use cases where we want to display the actual IDs in the UI. For such cases, I would like to fetch the original ID from the query.

**Reply by zzzeek** [reply]
the wiki example in my previous comment illustrates a fully transparent server side encrypt/decrypt function for use with the ORM
